python_modules/dagster/dagster/_core/definitions/external_asset.py

```diff
@@ -184,12 +186,44 @@ def _shim_assets_def(context: AssetExecutionContext):
     return _shim_assets_def
 
 
-# Create unexecutable assets defs for each asset key in the provided assets def. This is used to
-# make a materializable assets def available only for loading in a job.
-def create_unexecutable_external_assets_from_assets_def(
+# AssetsDefinition.get_io_manager_key_for_asset_key assumes an output is defined for the asset key,
+# which is not true if specs were used. Return None on the error case (no defined output).
+def _get_io_manager_key_for_unexecutable_asset(assets_def: AssetsDefinition, asset_key: AssetKey):
+    try:
+        return assets_def.get_io_manager_key_for_asset_key(asset_key)
+    except DagsterInvariantViolationError:
+        return None
```

Comment:
i think it would be good to detect this case in some other way vs. having a blanket "except DagsterInvariantViolationError" - i worry that this will potentially mask other errors, particularly if it does not log anything

Reply:
Perhaps even a specialized `DagsterInvariantViolationError` subclass so that the except clause is very precise.

Reply:
I've changed this so that we're no longer catching an error, but instead performing a check before the fetch